File: Code/Ifly/PresenterModeConfiguration.cs
namespace Ifly
{
    /// <summary>
    /// Represents presenter mode animation availability.
    /// </summary>
    public enum PresenterModeAnimationAvailability
    {
        /// <summary>
        /// Minimal transitions.
        /// </summary>
        Minimal = 0,

        /// <summary>
        /// Normal transitions.
        /// </summary>
        Normal = 1,

        /// <summary>
        /// No transitions.
        /// </summary>
        None = 2
    }

    /// <summary>
    /// Represents presenter mode configuration.
    /// </summary>
    public class PresenterModeConfiguration
    {
        /// <summary>
        /// Gets or sets the Id of the related presentation.
        /// </summary>
        public int PresentationId { get; set; }

        /// <summary>
        /// Gets or sets the availability of the element/slide animations.
        /// </summary>
        public PresenterModeAnimationAvailability Animations { get; set; }

        /// <summary>
        /// Gets or sets value indicating whether Sprites can ask to go fullscreen.
        /// </summary>
        public bool? AllowFullscreen { get; set; }

        /// <summary>
        /// Initializes a new instance of an object.
        /// </summary>
        public PresenterModeConfiguration()
        {
            Animations = PresenterModeAnimationAvailability.Minimal;
            AllowFullscreen = true;
        }

        /// <summary>
        /// Returns the absolute URI of a given presentation.
        /// </summary>
        /// <param name="requestUri">Request URI.</param>
        /// <param name="config">Presenter configuration.</param>
        /// <returns>The absolute URI of a given presentation.</returns>
        /// <exception cref="System.ArgumentNullException"><paramref name="requestUri" /> or <paramref name="config" /> is null.</exception>
        public static string GetAbsoluteUri(System.Uri requestUri, PresenterModeConfiguration config)
        {
            if (requestUri == null)
                throw new System.ArgumentNullException("requestUri");

            if (config == null)
                throw new System.ArgumentNullException("config");

            return GetAbsoluteUri(requestUri, config.PresentationId);
        }

        /// <summary>
        /// Returns the absolute URI of a given presentation.
        /// </summary>
        /// <param name="requestUri">Request URI.</param>
        /// <param name="presentationId">Presentation Id.</param>
        /// <returns>The absolute URI of a given presentation.</returns>
        /// <exception cref="System.ArgumentNullException"><paramref name="requestUri" /> is null.</exception>
        public static string GetAbsoluteUri(System.Uri requestUri, int presentationId)
        {
            string ret = PublishConfiguration.GetAbsoluteUri(requestUri, presentationId);

            ret += ((ret.IndexOf('?') > 0 ? '&' : '?') + "presenter=1");

            return ret;
        }

        /// <summary>
        /// Returns the Id of the presentation from the given URL.
        /// </summary>
        /// <param name="requestUri">Request URI.</param>
        /// <returns>Presentation Id.</returns>
        public static int GetPresentationIdFromUri(System.Uri requestUri)
        {
            return PublishConfiguration.GetPresentationIdFromUri(requestUri);
        }
    }
}

File: Code/Ifly.Web.Editor/Assets/js/Models/PresenterModeConfiguration.ts
/// <reference path="../Typings/knockout.d.ts" />
/// <reference path="IModel.ts" />

module Ifly.Models {
    /** Represents a presenter mode configuration. */
    export class PresenterModeConfiguration implements IModel {
        /** Gets or sets the Id of the related presentation. */
        public presentationId: KnockoutObservable<number>;

        /** Gets or sets the availability of the element/slide animations. */
        public animations: KnockoutObservable<PresenterModeAnimationAvailability>;

        /** Gets or sets value indicating whether Sprites is allowed to ask for fullscreen mode. */
        public allowFullscreen: KnockoutObservable<boolean>;

        /** Gets the absolute URL of the infographic. */
        public url: KnockoutObservable<string>;

        /** 
         * Initializes a new instance of an object.
         * @param {object} data Data to load from.
         */
        constructor(data?: any) {
            this.presentationId = ko.observable<number>();
            this.animations = ko.observable<PresenterModeAnimationAvailability>();
            this.allowFullscreen = ko.observable<boolean>();
            this.url = ko.observable<string>();
        }

        /** 
         * Populates object state from the given data.
         * @param {object} data Data to load from.
         */
        public load(data: any) {
            var ifDefined = v => {
                return typeof (v) !== 'undefined' && v !== null;
            };

            data = data || {};

            this.presentationId(data.presentationId || data.PresentationId || -1);
            this.animations(this.valueOrDefault<PresenterModeAnimationAvailability>(data, 'animations', 'Animations', PresenterModeAnimationAvailability.minimal));
            this.allowFullscreen(this.valueOrDefault<boolean>(data, 'allowFullscreen', 'AllowFullscreen', true));
            this.url(data.url || data.Url || '');
        }

        /** Serializes object state. */
        public serialize() {
            return {
                presentationId: this.presentationId(),
                animations: this.animations(),
                allowFullscreen: this.allowFullscreen(),
                url: this.url()
            };
        }

        /** 
         * Returns either the specified value or the default one.
         * @param {object} data Data object.
         * @param {string} p1 Property #1 probe.
         * @param {string} p2 Property #2 probe.
         * @param {T} defautValue Default value.
         */
        private valueOrDefault<T>(data: any, p1: string, p2: string, defaultValue: T): T {
            var ret = defaultValue;

            if (typeof (data[p1]) != 'undefined' || typeof (data[p2]) != 'undefined') {
                ret = typeof (data[p1]) != 'undefined' ? data[p1] : data[p2];
            }

            return ret;
        }
    }
} 